src/main/java/com/shapesecurity/shift/es2017/parser/Tokenizer.java

```diff
@@ -699,31 +692,42 @@ private CharSequence getIdentifier() throws JsError {
             if (ch == '\\' || 0xD800 <= code && code <= 0xDBFF) {
                 // Go back and try the hard one.
                 this.index = start;
-                return this.getEscapedIdentifier();
+                return Pair.of(this.getEscapedIdentifier(), true);
             }
             if (!check.apply(code)) {
                 this.index = i;
-                return this.source.subSequence(start, i);
+                return Pair.of(this.source.subSequence(start, i), false);
             }
             ++i;
             check = Utils::isIdentifierPart;
         }
         this.index = i;
-        return this.source.subSequence(start, i);
+        return Pair.of(this.source.subSequence(start, i), false);
     }
 
     @Nonnull
     private Token scanIdentifier() throws JsError {
         int start = this.index;
 
+        boolean escaped = this.source.charAt(this.index) == '\\';
+
         // Backslash (U+005C) starts an escaped character.
-        CharSequence id = this.source.charAt(this.index) == '\\' ? this.getEscapedIdentifier() : this.getIdentifier();
+        CharSequence id;
+        if (escaped) {
+            id = this.getEscapedIdentifier();
+        } else {
+            Pair<CharSequence, Boolean> pair = this.getIdentifier();
+            id = pair.left;
+            escaped = pair.right;
+        }
 
         SourceRange slice = this.getSlice(start);
 
         TokenType subType = this.getKeyword(id);
         if (subType == TokenType.IDENTIFIER) {
-            return new IdentifierToken(slice, id);
+            return new IdentifierToken(subType, slice, id, escaped);
+        } else if (escaped) {
+            return new IdentifierToken(TokenType.ESCAPED_KEYWORD, slice, id, true);
```

Comment:
Ditto above comment, actually - why is this an `IdentifierToken` instead of a `KeywordToken` with subtype `ESCAPED_KEYWORD`?

Reply:
The new parameter is used for escaped identifiers (used for contextual keywords), while `ESCAPED_KEYWORD` is used for known keywords that are escaped, so the parameter will still be needed. I can change it to a KeywordToken with some other changes.